modules/auxiliary/scanner/scada/bacnet_l3.rb

```diff
@@ -0,0 +1,412 @@
+      asset_data['sadr'] = bin_to_hex(items[3]) if items[0] == true
+      type = data[index + 8..index + 9]
+      attribute = ''
+      case type
+      when READ_MODEL_NAME_PROP
+        attribute = 'model-name'
+      when READ_DESCRIPTION_PROP
+        attribute = 'description'
+      when READ_APP_SOFT_VERSION_PROP
+        attribute = 'application-software-version'
+      when READ_FIRMWARE_VERSION_PROP
+        attribute = 'firmware-revision'
+      else
+        raise "undefined attribute for property number #{bin_to_hex(type)}."
+      end
+      begin
+        value = bin_to_hex(data[index + 9..])[/3e(.*?)3f/m, 1]
+        value = hex_to_bin(value)
+        value = (value[value.index(hex_to_bin('00')) + 1..]).force_encoding('UTF-8') # parsing the needed text
+        asset_data[attribute] = value
+      rescue StandardError => e
+        raise e
+      end
+    end
+    asset_data
+  end
+
+  # Gets properties from devices and returns a hash with the details of each device.
+  def get_properties_from_devices(messages, udp_pkt)
+    devices_by_ip = {}
+    messages.each do |key, message_block|
+      instance_number = key[0].to_i(16)
+      mac = key[1]
+      ip = key[2]
+
+      capture = send_read_properties(message_block, udp_pkt, mac, ip, instance_number)
+      begin
+        device = extract_data(capture)
+        raise StandardError if device.empty?
+
+        device['instance-number'] = instance_number.to_s
+        devices_by_ip[ip] = [] unless devices_by_ip[ip]
+        devices_by_ip[ip].append(device)
+      rescue StandardError
+        print_bad("Couldn't collect data for asset number #{instance_number}.")
+      end
+    end
+    devices_by_ip
+  end
+
+  # Sending read-property packets and returns a pcap with the responses.
+  def send_read_properties(messages, udp_pkt, mac, ip, instance_number)
+    udp_pkt.eth_daddr = mac
+    udp_pkt.ip_daddr = ip
+
+    capture = PacketFu::Capture.new(iface: $iface, start: true,
+                                    filter: "ip src #{ip} and ip dst #{udp_pkt.ip_saddr}")
+    print_status("Querying device number #{instance_number} in ip #{ip}")
+    messages.each do |message|
+      udp_pkt.payload = message
+      udp_pkt.recalc
+      $packet_count.times { udp_pkt.to_w($iface) }
+    end
+    sleep($timeout)
+    capture.save
+    capture
+  end
+
+  # Iterates over all the devices and prints the details to the user.
+  def output_results(devices_by_ip)
+    devices_by_ip.each_value do |ip_group|
+      ip_group.each do |asset|
+        sadr = ''
+        if asset['sadr']
+          sadr = "sadr: #{asset['sadr']}
+              "
+        end
+        print_good "for asset number #{asset['instance-number']}:
+          model name: #{asset['model-name']}
+          firmware revision: #{asset['firmware-revision']}
+          application software version: #{asset['application-software-version']}
+          description: #{asset['description']}
+          #{sadr}"
```

Comment:
You can use heredocs for better control over the output format. For example (untested):
```suggestion
        print_good(<<~OUTPUT)
          for asset number #{asset['instance-number']}:
          \tmodel name: #{asset['model-name']}
          \tfirmware revision: #{asset['firmware-revision']}
          \tapplication software version: #{asset['application-software-version']}
          \tdescription: #{asset['description']}
          \t#{sadr}
          OUTPUT
```

Reply:
I also noticed there are other locations in the code where heredocs would be a better options for multi-line strings.